### PR TITLE
feat(VENLY-3557) IMX Balance gateway

### PR DESCRIPTION
**Pull request by maciejwitwicki**

### CODE REVIEW

File: provider-api-model/src/main/java/network/arkane/provider/balance/domain/Balance.java [hunk 1]
```diff
@@ -1,10 +1,15 @@
...
+import lombok.AllArgsConstructor;
 import lombok.Builder;
 import lombok.Data;
+import lombok.NoArgsConstructor;
 import network.arkane.provider.chain.SecretType;
 
 @Data
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder(toBuilder = true)
```
Comment (aleksanderndoci) [inline provider-api-model/src/main/java/network/arkane/provider/balance/domain/Balance.java:12]: Do you need the toBuilder ?

File: provider-api/src/main/java/network/arkane/provider/balance/BalanceBlockchainProvider.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+import org.springframework.http.client.support.BasicAuthenticationInterceptor;
+import org.springframework.web.client.RestTemplate;
+import org.springframework.web.util.DefaultUriBuilderFactory;
+
+import java.time.Duration;
+import java.util.List;
+
+import static java.time.Duration.ofSeconds;
+
+public class BalanceBlockchainProvider extends BalanceGateway {
```
Comment (aleksanderndoci) [inline provider-api/src/main/java/network/arkane/provider/balance/BalanceBlockchainProvider.java:19]: I would name this BlockchainProviderBalanceGateway

File: provider-imx/src/main/java/io/venly/provider/imx/ImxAutoConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@EnableConfigurationProperties({ImxProperties.class})
+public class ImxAutoConfiguration {
+
+    @PostConstruct
+    public void init() {
+        log.info("provider-imx loaded");
+    }
+
+    @Bean
+    @ConditionalOnProperty(prefix = "io.venly.provider.imx", name = {"endpoint", "user", "password"})
```
Comment (aleksanderndoci) [inline provider-imx/src/main/java/io/venly/provider/imx/ImxAutoConfiguration.java:25]: Wouldn't it be simpler to mark ImxGateway with @Component and add the @ConditionalOnProperty there?

File: provider-imx/src/main/resources/META-INF/spring.factories [hunk 4]
```diff
@@ -0,0 +1 @@
+org.springframework.boot.autoconfigure.EnableAutoConfiguration=io.venly.provider.imx.ImxAutoConfiguration
```
Comment (aleksanderndoci) [inline provider-imx/src/main/resources/META-INF/spring.factories:1]: I don't think this is needed. There is already an AutoConfiguration done for the whole project in `provider-autoconfiguration`

File: provider-imx/src/main/java/io/venly/provider/imx/config/ImxProperties.java [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package io.venly.provider.imx.config;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties(prefix = "io.venly.provider.imx")
```
Comment (aleksanderndoci) [inline provider-imx/src/main/java/io/venly/provider/imx/config/ImxProperties.java:5]: I would consider renaming this to something that includes `imx-gateway` so it's clear where it points to

File: provider-api/pom.xml [hunk 6]
```diff
@@ -72,6 +72,10 @@
             <artifactId>lombok</artifactId>
             <scope>provided</scope>
         </dependency>
+        <dependency>
```
Comment (aleksanderndoci) [inline provider-api/pom.xml:75]: We are using MockMvc across the whole projects. I would suggest continuing to use that unless there's a specific reason you're using wiremock

File: provider-api/src/test/java/network/arkane/provider/balance/BalanceBlockchainProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import static com.github.tomakehurst.wiremock.client.WireMock.verify;
+import static org.apache.commons.codec.binary.Base64.encodeBase64String;
+import static org.assertj.core.api.Assertions.assertThat;
+
+@WireMockTest
+class BalanceBlockchainProviderTest {
+
+    @Test
+    void shouldCallBalanceEndpoint(WireMockRuntimeInfo runtime) {
+        final String user = "user";
```
Comment (aleksanderndoci) [inline provider-api/src/test/java/network/arkane/provider/balance/BalanceBlockchainProviderTest.java:31]: Can MockMvc be used instead of wiremock ?

**Review by aleksanderndoci — CHANGES_REQUESTED** [review]

File: provider-imx/src/main/java/io/venly/provider/imx/ImxAutoConfiguration.java (hunk 3 above)
Comment (maciejwitwicki) [reply provider-imx/src/main/java/io/venly/provider/imx/ImxAutoConfiguration.java:25]: @Component would not be picked by scanner alone, we would then need to point the package in the project that uses it.

Comment (aleksanderndoci) [reply provider-imx/src/main/java/io/venly/provider/imx/ImxAutoConfiguration.java:25]: Hmm, but there is a provider-autoconfiguration module which allows component scanning from package `network.arkane.provider.**`  which I think covers all the packages on deeper level

File: provider-imx/src/main/resources/META-INF/spring.factories (hunk 4 above)
Comment (maciejwitwicki) [reply provider-imx/src/main/resources/META-INF/spring.factories:1]: As I have migrated the package to io.venly.provider the provider-autoconfiguration will miss the IMX provider without explicitly setting additional place to scan here:  [BlockchainAutoconfiguration.java](https://github.com/ArkaneNetwork/blockchain-providers/blob/d13fea7f109164706d06f623e60a8b36ddf5493f/provider-autoconfiguration/src/main/java/network/arkane/provider/BlockchainAutoconfiguration.java#L10)

Comment (aleksanderndoci) [reply provider-imx/src/main/resources/META-INF/spring.factories:1]: Yup that is true, I missed that part

File: provider-api/pom.xml (hunk 6 above)
Comment (maciejwitwicki) [reply provider-api/pom.xml:75]: I'll switch to MockMvc 👍

**Pull request merged**